java-checks/src/main/java/org/sonar/java/checks/VarCanBeUsedCheck.java

```diff
@@ -0,0 +1,95 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Collections;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.java.JavaVersionAwareVisitor;
+import org.sonar.java.model.JUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.SyntaxToken;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6212")
+public class VarCanBeUsedCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+  
+  private int typeAssignmentLine = -1;
+  
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.VARIABLE);
+  }
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava10Compatible();
+  }
+
+  @Override
+  public void setContext(JavaFileScannerContext context) {
+    typeAssignmentLine = -1;
+    super.setContext(context);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    VariableTree variableTree = (VariableTree) tree;
+    ExpressionTree initializer = variableTree.initializer();
+    TypeTree type = variableTree.type();
+    Type symbolType = type.symbolType();
+    
+    if (isMultiAssignment(variableTree) ||
+      initializer == null ||
+      type.is(Tree.Kind.VAR_TYPE) ||
+      symbolType.isUnknown() ||
+      (symbolType.isPrimitive() && !(initializer instanceof LiteralTree)) ||
+      !JUtils.isLocalVariable(variableTree.symbol()) ||
+      initializer.is(Tree.Kind.CONDITIONAL_EXPRESSION)) {
+      return;
+    }
+
+    Type initializerType = initializer.symbolType();
+    if (symbolType.fullyQualifiedName().equals(initializerType.fullyQualifiedName())) {
+      reportIssue(variableTree.simpleName(), "");
+    }
+  }
+
+  private boolean isMultiAssignment(VariableTree variableTree) {
+    SyntaxToken firstToken = variableTree.type().firstToken();
+    if (firstToken == null) {
+      return false;
+    }
+    int line = firstToken.line();
+    if (typeAssignmentLine == line) {
+      return true;
+    }
+    typeAssignmentLine = line;
+    SyntaxToken token = variableTree.endToken();
+    return token != null && token.text().equals(",");
+  }
```

Comment:
This code looks strange but I'm not sure there is another way to write it... So I guess we will keep it as it is.

Reply:
I tried to find how it's handled in other rules and it was a similar approach. In situations like this `int a = 0, b = 1;` we can't put `var`, but there is no dedicated Kind. Maybe we should have it in the future.